File: packages/gavio-js/src/interceptors/cache/backends/redis.ts
/**
 * Redis cache backends (F-CACHE-04) — production-grade distributed cache.
 *
 * Zero runtime dependencies — talks RESP2 directly over `node:net` (see
 * `./resp.js`). In-memory backends remain the zero-infra default.
 */

import { randomUUID } from 'node:crypto'
import type { CacheBackend } from '../backend.js'
import { cosineSimilarity } from '../embedding.js'
import type { VectorBackend } from '../vector.js'
import { parseRedisUrl, RespClient, type RespValue } from './resp.js'

export interface RedisCacheBackendOptions {
  url?: string
  namespace?: string
}

function asStringArray(value: RespValue): string[] {
  if (!Array.isArray(value)) return []
  return value.filter((v): v is string => typeof v === 'string')
}

/** Exact-match `CacheBackend` over Redis. Keys are namespaced so `clear()` never
 * touches entries another backend/namespace wrote. */
export function redisCacheBackend(options: RedisCacheBackendOptions = {}): CacheBackend {
  const { host, port } = parseRedisUrl(options.url ?? 'redis://localhost:6379')
  const namespace = options.namespace ?? 'gavio:cache'
  const prefix = `${namespace}:`
  const indexKey = `${namespace}:index`
  const client = new RespClient(host, port)
  const namespaced = (key: string): string => prefix + key

  return {
    async get(key: string): Promise<unknown | null> {
      const raw = await client.command('GET', namespaced(key))
      if (typeof raw !== 'string') {
        await client.command('SREM', indexKey, key)
        return null
      }
      return JSON.parse(raw) as unknown
    },
    async set(key: string, value: unknown, ttlSeconds?: number | null): Promise<void> {
      const raw = JSON.stringify(value)
      if (ttlSeconds) await client.command('SET', namespaced(key), raw, 'EX', ttlSeconds)
      else await client.command('SET', namespaced(key), raw)
      await client.command('SADD', indexKey, key)
    },
    async delete(key: string): Promise<void> {
      await client.command('DEL', namespaced(key))
      await client.command('SREM', indexKey, key)
    },
    async clear(): Promise<void> {
      const keys = asStringArray(await client.command('SMEMBERS', indexKey))
      if (keys.length > 0) await client.command('DEL', ...keys.map(namespaced))
      await client.command('DEL', indexKey)
    },
  }
}

export interface RedisVectorBackendOptions {
  url?: string
  namespace?: string
}

/** Brute-force cosine-similarity `VectorBackend` over Redis — same matching
 * strategy as `inMemoryVectorBackend`, just shared across processes. */
export function redisVectorBackend(options: RedisVectorBackendOptions = {}): VectorBackend {
  const { host, port } = parseRedisUrl(options.url ?? 'redis://localhost:6379')
  const namespace = options.namespace ?? 'gavio:vector'
  const indexKey = `${namespace}:index`
  const client = new RespClient(host, port)
  const entryKey = (id: string): string => `${namespace}:${id}`

  return {
    async add(vector: number[], value: unknown, ttlSeconds?: number | null): Promise<void> {
      const id = randomUUID()
      const raw = JSON.stringify({ vector, value })
      if (ttlSeconds) await client.command('SET', entryKey(id), raw, 'EX', ttlSeconds)
      else await client.command('SET', entryKey(id), raw)
      await client.command('SADD', indexKey, id)
    },
    async query(vector: number[], threshold: number): Promise<unknown | null> {
      const ids = asStringArray(await client.command('SMEMBERS', indexKey))
      let best: unknown = null
      let bestSim = threshold
      for (const id of ids) {
        const raw = await client.command('GET', entryKey(id))
        if (typeof raw !== 'string') {
          await client.command('SREM', indexKey, id)
          continue
        }
        const entry = JSON.parse(raw) as { vector: number[]; value: unknown }
        const sim = cosineSimilarity(vector, entry.vector)
        if (sim >= bestSim) {
          bestSim = sim
          best = entry.value
        }
      }
      return best
    },
    async clear(): Promise<void> {
      const ids = asStringArray(await client.command('SMEMBERS', indexKey))
      if (ids.length > 0) await client.command('DEL', ...ids.map(entryKey))
      await client.command('DEL', indexKey)
    },
  }
}

File: packages/gavio-py/gavio/interceptors/cache/backends/redis.py
"""Redis cache backends (F-CACHE-04) — production-grade distributed cache.

Optional — requires ``pip install gavio[redis]``. The in-memory backends
remain the zero-infra default (design principle P4); this module is only
imported when a caller actually constructs ``RedisBackend``/``RedisVectorBackend``.
"""

from __future__ import annotations

import json
import uuid
from typing import Any

from ..backend import CacheBackend
from ..embedding import cosine_similarity
from ..vector import VectorBackend

try:
    from redis.asyncio import Redis as _AsyncRedis
except ImportError:  # pragma: no cover - exercised only without the extra installed
    _AsyncRedis = None  # type: ignore[assignment,misc]


def _require_redis() -> None:
    if _AsyncRedis is None:
        raise ImportError(
            "RedisBackend requires the 'redis' package — install with `pip install gavio[redis]`"
        )


class RedisBackend(CacheBackend):
    """Exact-match ``CacheBackend`` over Redis (F-CACHE-04).

    Keys are namespaced under an index set so ``clear()`` only removes entries
    this backend itself wrote, never the whole database.
    """

    def __init__(
        self,
        url: str = "redis://localhost:6379",
        namespace: str = "gavio:cache",
        client: Any | None = None,
    ) -> None:
        _require_redis()
        self._client = client or _AsyncRedis.from_url(url, decode_responses=True)
        self._prefix = f"{namespace}:"
        self._index_key = f"{namespace}:index"

    def _namespaced(self, key: str) -> str:
        return self._prefix + key

    async def get(self, key: str) -> Any | None:
        raw = await self._client.get(self._namespaced(key))
        if raw is None:
            await self._client.srem(self._index_key, key)
            return None
        return json.loads(raw)

    async def set(self, key: str, value: Any, ttl_seconds: int | None = None) -> None:
        raw = json.dumps(value)
        if ttl_seconds:
            await self._client.set(self._namespaced(key), raw, ex=ttl_seconds)
        else:
            await self._client.set(self._namespaced(key), raw)
        await self._client.sadd(self._index_key, key)

    async def delete(self, key: str) -> None:
        await self._client.delete(self._namespaced(key))
        await self._client.srem(self._index_key, key)

    async def clear(self) -> None:
        keys = await self._client.smembers(self._index_key)
        if keys:
            await self._client.delete(*(self._namespaced(str(k)) for k in keys))
        await self._client.delete(self._index_key)


class RedisVectorBackend(VectorBackend):
    """Brute-force cosine-similarity ``VectorBackend`` over Redis (F-CACHE-04).

    Same brute-force matching strategy as ``InMemoryVectorBackend`` — just
    shared across processes. Fine for the cache's scale (bounded, TTL'd
    entries); not a substitute for a real vector database.
    """

    def __init__(
        self,
        url: str = "redis://localhost:6379",
        namespace: str = "gavio:vector",
        client: Any | None = None,
    ) -> None:
        _require_redis()
        self._client = client or _AsyncRedis.from_url(url, decode_responses=True)
        self._namespace = namespace
        self._index_key = f"{namespace}:index"

    async def add(
        self, vector: list[float], value: Any, ttl_seconds: int | None = None
    ) -> None:
        entry_id = uuid.uuid4().hex
        key = f"{self._namespace}:{entry_id}"
        raw = json.dumps({"vector": vector, "value": value})
        if ttl_seconds:
            await self._client.set(key, raw, ex=ttl_seconds)
        else:
            await self._client.set(key, raw)
        await self._client.sadd(self._index_key, entry_id)

    async def query(self, vector: list[float], threshold: float) -> Any | None:
        entry_ids = await self._client.smembers(self._index_key)
        best_value: Any | None = None
        best_sim = threshold
        for entry_id in entry_ids:
            entry_id = str(entry_id)
            key = f"{self._namespace}:{entry_id}"
            raw = await self._client.get(key)
            if raw is None:
                await self._client.srem(self._index_key, entry_id)
                continue
            entry = json.loads(raw)
            sim = cosine_similarity(vector, entry["vector"])
            if sim >= best_sim:
                best_sim = sim
                best_value = entry["value"]
        return best_value

    async def clear(self) -> None:
        entry_ids = await self._client.smembers(self._index_key)
        keys = [f"{self._namespace}:{entry_id!s}" for entry_id in entry_ids]
        if keys:
            await self._client.delete(*keys)
        await self._client.delete(self._index_key)
